### PR TITLE
[SPARK-34781][SQL] Eliminate LEFT SEMI/ANTI joins to its left child side in AQE

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

In `EliminateJoinToEmptyRelation.scala`, we can extend it to cover more cases for LEFT SEMI and LEFT ANI joins:

* Join is left semi join, join right side is non-empty and condition is empty. Eliminate join to its left side.
* Join is left anti join, join right side is empty. Eliminate join to its left side.

Given we eliminate join to its left side here, renaming the current optimization rule to `EliminateUnnecessaryJoin` instead.
In addition, also change to use `checkRowCount()` to check run time row count, instead of using `EmptyHashedRelation`. So this can cover `BroadcastNestedLoopJoin` as well. (`BroadcastNestedLoopJoin`'s broadcast side is `Array[InternalRow]`, not `HashedRelation`).

### Why are the changes needed?

Cover more join cases, and improve query performance for affected queries.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unit tests in `AdaptiveQueryExecSuite.scala`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+object EliminateUnnecessaryJoin extends Rule[LogicalPlan] {
+
+  private def isRelationWithAllNullKeys(plan: LogicalPlan) = plan match {
+    case LogicalQueryStage(_, stage: BroadcastQueryStageExec)
+      if stage.resultOption.get().isDefined =>
+      stage.broadcast.relationFuture.get().value == HashedRelationWithAllNullKeys
+    case _ => false
+  }
+
+  private def checkRowCount(plan: LogicalPlan, hasRow: Boolean): Boolean = plan match {
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:56]: This is added compared to previous `EliminateJoinToEmptyRelation.scala`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ *    4.2. Join right side is non-empty and condition is empty. Eliminate join to an empty
+ *         [[LocalRelation]].
+ *
+ * This applies to all joins (sort merge join, shuffled hash join, broadcast hash join, and
+ * broadcast nested loop join), because sort merge join and shuffled hash join will be changed
+ * to broadcast hash join with AQE at the first place.
+ */
+object EliminateUnnecessaryJoin extends Rule[LogicalPlan] {
+
+  private def isRelationWithAllNullKeys(plan: LogicalPlan) = plan match {
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:49]: This is slightly changed compared to previous `EliminateJoinToEmptyRelation.scala`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 3]
```diff
@@ -1254,21 +1254,38 @@ class AdaptiveQueryExecSuite
...
-          val (plan, adaptivePlan) = runAdaptiveAndVerifyResult(query)
-          assert(findTopLevelBaseJoin(plan).size == 1)
-          assert(findTopLevelBaseJoin(adaptivePlan).isEmpty == isEliminated)
-        }
+      Seq(
+        // broadcast non-empty right side
+        ("SELECT /*+ broadcast(testData3) */ * FROM testData LEFT ANTI JOIN testData3", true),
+        // broadcast empty right side
+        ("SELECT /*+ broadcast(emptyTestData) */ * FROM testData LEFT ANTI JOIN emptyTestData",
+          true),
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1262]: This is true now as we eliminate join to its left side.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 4]
```diff
@@ -1254,21 +1254,38 @@ class AdaptiveQueryExecSuite
   test("SPARK-34533: Eliminate left anti join to empty relation") {
     withSQLConf(
       SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true") {
-      withTable("emptyTestData") {
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1257]: Do not create `emptyTestData` anymore as we already [created one here](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestData.scala#L295). Create table again here will cause problem for the followed newly added test when trying to use `emptyTestData`.

**Comment by c21** [comment]
cc @cloud-fan and @maropu could you help take a look when you have time, thanks.

**Comment by c21** [comment]
retest this please

**Comment by c21** [comment]
Closed & reopened the PR to trigger test rerun.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:56]: nit comment: we need the `hasRow` param? How about simply checking `count == 0` like this?
```
private def hasNoRow(plan: LogicalPlan): Boolean = plan match {
    case LogicalQueryStage(_, stage: QueryStageExec) if stage.resultOption.get().isDefined =>
      stage.getRuntimeStatistics.rowCount match {
        case Some(count) => count == 0
        case _ => false
      }
    case _ => false
  }
```
Then,
```
    case j @ Join(_, _, Inner, _, _) if hasNoRow(j.left) ||
    ...
    case j @ Join(_, _, LeftSemi, condition, _) =>
      if (checkRowCount(j.right, hasRow = false)) {
        LocalRelation(j.output, data = Seq.empty, isStreaming = j.isStreaming)
      } else if (condition.isEmpty && !hasNoRow(j.right)) {
```

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
I left the single minor comment and it looks fine otherwise.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:56]: @maropu - sorry I don't quite get here, `checkRowCount(j.right, hasRow = false` still exists in your example. Am I missing anything? Thanks.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:56]: ah, my bad. I meant `checkRowCount` -> `hasNoRow `:
```
    case j @ Join(_, _, Inner, _, _) if hasNoRow(j.left) ||
    ...
    case j @ Join(_, _, LeftSemi, condition, _) =>
      if (hasNoRow(j.right)) {
        LocalRelation(j.output, data = Seq.empty, isStreaming = j.isStreaming)
      } else if (condition.isEmpty && !hasNoRow(j.right)) {
```

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:56]: @maropu - I thought about this too, but I think this is dangerous as we want to make sure the child output is known, and row count is larger than zero.

There can be the case where `hasNoRow()` returns false but the plan is not executed in AQE at all (not `LogicalQueryStage`) or result is not known yet.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala:56]: Ah, okay. I got your point. sgtm

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by c21** [comment]
Thank you @maropu and @cloud-fan for review!

**Pull request closed without merging**